### DISCUSSION TITLE
Latest version `.predict` is much slower than version 8.3.181

### DISCUSSION THREAD

**Discussion by tonyreina** [ultralytics/ultralytics#22109] [General]
I tried updating my Ultralytics package to the latest version, but am finding it to be much slower than my original version 8.3.181.  I'm just using the YOLO11n model and the `model.predict`. My code is exactly the same. For version 8.3.181 I can get about 10 FPS. For the latest version 8.3.200, I am getting several seconds per frame and it gives me "WARNING:  NMS time limit 2.800s exceeded".

Can anyone think of a way to diagnose what is slowing things down?  I'm assuming the NMS message has something to do with it. I don't get any such method for Ultralytics 8.3.181.

ultralytics                               8.3.181                                         2.9 MiB     pypi   ultralytics-8.3.181-py3-none-any.whl
ultralytics_thop                      2.0.15                                          82.2 KiB     pypi   ultralytics_thop-2.0.15-py3-none-any.whl

Thanks.
-Tony

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tonyreina, thanks for reaching out and for testing the latest Ultralytics release 🚀 This is an automated response to help get things started. An Ultralytics engineer will also assist you here shortly.

Please check out the [Docs](https://docs.ultralytics.com/) for quick references on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, as many common questions are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the performance difference you’re seeing. Since you mentioned a regression between 8.3.181 and 8.3.200 with an NMS warning, it would be very helpful if you could include:
- A small script that reproduces the slowdown with `YOLO11n` and `model.predict()`
- 1–2 sample images/videos you’re running inference on
- Exact command-line or Python args (e.g., imgsz, conf, iou, device, half, stream, save, verbose)
- Environment details: OS, GPU, driver, CUDA/CUDNN, Python, PyTorch, and ultralytics versions
- Any logs showing the “NMS time limit exceeded” message and timing per step if available

If this is a custom training ❓ Question, please include dataset samples, training logs, and verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify the issue persists on the newest release:

```bash
pip install -U ultralytics
```

If you can, try a clean virtual environment and re-test both versions to compare apples-to-apples. Sharing your side-by-side timings will help a lot ⚙️

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

These can help confirm whether the slowdown is environment-specific.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks for the detailed version info you’ve already shared. With an MRE plus environment details and logs, we’ll be able to diagnose the NMS timing behavior more quickly 🔍

**Comment by tonyreina** [comment]
I checked the versions and the code slows down between versions 8.3.189 and 8.3.190.

I'm thinking that the new NMS code is the culprit.

**Comment by tonyreina** [comment]
https://github.com/ultralytics/ultralytics/compare/v8.3.189...v8.3.190#diff-f9841f219d5e5a587d53e796e229175de2dee78023c1ed3ec1d3754054d278a6

**Comment by tonyreina** [comment]
I was able to monkey patch it to use the old NMS code. It might be nice to have a switch or something instead.

```

import torch
from torchvision.ops import nms as tv_nms

# 1) Patch the legacy/function-style hook if present
try:
    from ultralytics.utils import ops as uops

    if hasattr(uops, "nms"):
        # Expect signature (boxes[N,4], scores[N], iou_thres) -> idx (int64)
        uops.nms = tv_nms
        print("[patch] ultralytics.utils.ops.nms -> torchvision.ops.nms")
except Exception as e:
    print(f"[patch] ops.nms not patched: {e}")

# 2) Patch the newer class-style TorchNMS if present
try:
    from ultralytics.utils import ops as uops

    if hasattr(uops, "TorchNMS"):
        _orig_cls = uops.TorchNMS

        class TorchVisionNMS(_orig_cls):
            """Drop-in replacement that calls torchvision.ops.nms under the hood."""

            def __call__(
                self, boxes: torch.Tensor, scores: torch.Tensor, iou_thres: float
            ):
                # Ultralytics expects int64 indices on the same device
                # torchvision.ops.nms already does that efficiently.
                return tv_nms(boxes, scores, iou_thres)

        uops.TorchNMS = TorchVisionNMS
        print("[patch] ultralytics.utils.ops.TorchNMS -> TorchVisionNMS (tv.ops.nms)")
except Exception as e:
    print(f"[patch] TorchNMS not patched: {e}")
```

**Comment by Y-T-G** [comment]
It should use the `torchvision` NMS, if torchvision is already imported. Otherwise, it uses the Ultralytics implementation.

https://github.com/ultralytics/ultralytics/blob/dbc5433bc8058e28ecea422c7c2210a9acae7a32/ultralytics/utils/nms.py#L155

So you can add `import torchvision` in your code before making prediction.

**Reply by glenn-jocher** [reply]
Yes @tonyreina you can do:

```python
import torchvision

# rest of your code
```

This will detect the import and automatically use torchvision NMS later in your code.

**Comment by tonyreina** [comment]
That works. Thanks.

In my case I am using very large images (5120 x 5120) and the number of detection boxes is huge. I think that is why the new NMS is so slow compared to the old one.

It would be great if the `import torchvision` include were highlighted in the documentation.

**Reply by glenn-jocher** [reply]
Thanks for confirming, Tony. For very large frames with many candidates, you can cut NMS load by tightening thresholds and limiting classes/detections, e.g.:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
results = model.predict(
    source="big.jpg",
    conf=0.5,           # raise to reduce candidates
    iou=0.45,           # adjust as needed
    max_det=300,        # cap kept detections
    classes=[0, 1]      # only if you need a subset of classes
)
```

We’ll add a note in the docs that importing `torchvision` before inference switches to `torchvision.ops.nms`; see the NMS implementation and parameters in the [non_max_suppression reference](https://docs.ultralytics.com/reference/utils/nms/), and feel free to propose a small docs PR via the [contributing guide](https://docs.ultralytics.com/help/contributing/). If you’re testing in a notebook after upgrading, a kernel restart helps ensure the import path is picked up.

**Reply by cainiao33** [reply]
Following up here with measurements, since the docs note mentioned at the end hasn't landed yet.

On current main (8.4.142) the standard `predict()` path now preloads torchvision during `AutoBackend.warmup()` (#25023), so NMS runs on `torchvision.ops.nms` automatically. Verified end-to-end on Tony's scenario shape — a dense 5120×5120 frame (~537k anchors, ~12k candidates above conf): 0.38 s in a fresh process after a normal `predict()` call, vs 17.9 s when torchvision is never imported (CPU); on CUDA the same load goes 25.5 s → 0.037 s (AMD Ryzen 7 5700X, 8 torch threads, RTX 4060 Ti 16 GB, torch 2.14.0+cu130). In my tests the two implementations kept identical boxes and order.

The pure-torch fallback still affects two groups: releases between 8.3.190 and the one preceding #25023, and custom pipelines that call `non_max_suppression()` directly — for those, `import torchvision` before inference remains the workaround @Y-T-G described above.

@glenn-jocher happy to propose that small docs PR (a short note in the predict-mode FAQ) if still wanted — that would also cover @tonyreina's ask above.

**Reply by Y-T-G** [reply]
We don't highlight information about old versions in our docs. The docs always relay information about latest version. And the use case of calling NMS externally is very niche and not something we expect our users to do. It's an internal API and caveats with using internal API are not supposed to be part of docs

**Reply by cainiao33** [reply]
Understood — thanks for clarifying @Y-T-G. Closing the docs PR. The measurements above stay for anyone hitting this on affected versions; current-main users are covered by the warmup preload anyway.
